src/Shared/TaskLoggingHelper.cs

```diff
@@ -1016,6 +1043,9 @@ public void LogWarning
             // that gives the user something.
             bool fillInLocation = (String.IsNullOrEmpty(file) && (lineNumber == 0) && (columnNumber == 0));
 
+            // Keep track of warnings logged and compare to the what the build engine logged as an error.
+            _warningCodesLogged.Add(warningCode);
```

Comment:
Is this thread safe?

Reply:
Likely not, I can add a lock here.

Part of me wants to give every taskhost (on creation) the entire set of warnings as errors. Then I can make every task logger check their taskhost and, instead of logging a warning that will eventually be turned into an error, just have it log the error ahead of time.

That'd require:
1. Making sure every logger has access to a taskhost (or the set of warningsaserrors)
2. Ensuring every logger does this check when they want to log a warning
3. Ensuring this isn't a breaking change somehow
 
Not sure exactly how this would affect an outofproc task host. Any thoughts or suggestions?

Reply:
Yea, either lock (this class apparently already has a lock object) or concurrent collection.

Regarding the entire-set-of-warnings-for-each-taskhost approach, there's two benefits right?
- it works with out of proc task host whereas the current approach doesn't because the out of proc task host does not have access to a logging service.
- it's easier to reason about

---
In reply to: [562785330](https://github.com/dotnet/msbuild/pull/6040#discussion_r562785330) [](ancestors = 562785330)